File: tree/count_node_complete_tree.java
# concise idea from https://discuss.leetcode.com/topic/15533/concise-java-solutions-o-log-n-2
# author: Stefan Pochmann
class Solution {
    int height(TreeNode root) {
        return root == null ? -1 : 1 + height(root.left);
    }// same idea: count sub tree heights
    public int countNodes(TreeNode root) {
        int h = height(root);
        return h < 0 ? 0 :
               height(root.right) == h-1 ? (1 << h) + countNodes(root.right)
                                         : (1 << h-1) + countNodes(root.left);
    }// he is using binary operators 1<<h same as 2 to the power of h

}

File: graph(tree)/count_node_complete_tree.py
# 1. Straight forward sol'n
# if left sub tree height == right's
# then left sub tree is a perfect BT
# else, right is perfect
# return the count of perfect trees using power of 2
# count others recuisively
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def countNodes(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def depth(root):
            
            """
            :type root: TreeNode
            :rtype: int
            """
            if not root: return 0
            return 1+depth(root.left)
        
        if not root: return 0
        
        leftkid = depth(root.left)
        rightkid = depth(root.right)
        
        if leftkid == rightkid:
            # left sub tree perfect
            return pow(2, leftkid)+self.countNodes(root.right)
        else:
            # right sub tree perfect
            return pow(2, rightkid)+self.countNodes(root.left)

# LEARN USAGE OF NESTED FUNCTIONS IN PYTHON
